util: explicitly close all AutoFiles that have been written

vasild opened the pull request:
`fclose(3)` may fail to flush the previously written data to disk, thus a failing `fclose(3)` is as serious as a failing `fwrite(3)`.

Previously the code ignored `fclose(3)` failures. This PR improves that by changing all users of `AutoFile` that use it to write data to explicitly close the file and handle a possible error.

---

Other alternatives are:

1. `fflush(3)` after each write to the file (and throw if it fails from the `AutoFile::write()` method) and hope that `fclose(3)` will then always succeed. Assert that it succeeds from the destructor :roll_eyes:. Will hurt performance.
2. Throw nevertheless from the destructor. Exception within the exception in C++ I think results in terminating the program without a useful message.
3. (this is implemented in the latest incarnation of this PR) Redesign `AutoFile` so that its destructor cannot fail. Adjust _all_ its users :sob:. For example, if the file has been written to, then require the callers to explicitly call the `AutoFile::fclose()` method before the object goes out of scope. In the destructor, as a sanity check, assume/assert that this is indeed the case. Defeats the purpose of a RAII wrapper for `FILE*` which automatically closes the file when it goes out of scope and there are a lot of users of `AutoFile`.
4. Pass a new callback function to the `AutoFile` constructor which will be called from the destructor to handle `fclose()` errors, as described in https://github.com/bitcoin/bitcoin/pull/29307#issuecomment-2243842400. My thinking is that if that callback is going to only log a message, then we can log the message directly from the destructor without needing a callback. If the callback is going to do more complicated error handling then it is easier to do that at the call site by directly calling `AutoFile::fclose()` instead of getting the `AutoFile` object out of scope (so that its destructor is called) and inspecting for side effects done by the callback (e.g. set a variable to indicate a failed `fclose()`).

vasild committed:
Explicitly close all AutoFiles that have been written

There is no way to report a close error from `AutoFile` destructor.
Such an error could be serious if the file has been written to because
it may mean the file is now corrupted (same as if write fails).

So, change all users of `AutoFile` that use it to write data to
explicitly close the file and handle a possible error.

vasild commented on src/test/fuzz/autofile.cpp:
```diff
@@ -47,7 +47,7 @@ FUZZ_TARGET(autofile)
             },
             [&] {
-                auto_file.fclose();
+                assert(auto_file.fclose() == 0);
```
This assert actually failed in CI (wow!):

https://cirrus-ci.com/task/5586893237125120

```
Run autofile with args ['/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz', '-runs=1', PosixPath('/ci_container_base/ci/scratch/qa-assets/fuzz_seed_corpus/autofile')]INFO: Running with entropic power schedule (0xFF, 100).
INFO: Seed: 3765617163
INFO: Loaded 1 modules   (547059 inline 8-bit counters): 547059 [0x55cd54dd29b8, 0x55cd54e582ab), 
INFO: Loaded 1 PC tables (547059 PCs): 547059 [0x55cd54e582b0,0x55cd556b11e0), 
INFO:      535 files found in /ci_container_base/ci/scratch/qa-assets/fuzz_seed_corpus/autofile
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 1041296 bytes
INFO: seed corpus: files: 535 min: 1b max: 1041296b total: 17489928b rss: 106Mb
fuzz: test/fuzz/autofile.cpp:50: auto autofile_fuzz_target(FuzzBufferType)::(anonymous class)::operator()() const: Assertion `auto_file.fclose() == 0' failed.
==18378== ERROR: libFuzzer: deadly signal
    #0 0x55cd5193ab55 in __sanitizer_print_stack_trace (/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x1a5fb55) (BuildId: 34c8e9fbac2706ae466ade7d359adebec5d82bcf)
    #1 0x55cd518922fc in fuzzer::PrintStackTrace() (/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x19b72fc) (BuildId: 34c8e9fbac2706ae466ade7d359adebec5d82bcf)
    #2 0x55cd51878367 in fuzzer::Fuzzer::CrashCallback() (/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x199d367) (BuildId: 34c8e9fbac2706ae466ade7d359adebec5d82bcf)
    #3 0x7fb162f1d8ff  (/lib/x86_64-linux-gnu/libc.so.6+0x428ff) (BuildId: f0b834daa3d05a80967e9ec2f990a1ea71c958fa)
    #4 0x7fb162f7498a in pthread_kill (/lib/x86_64-linux-gnu/libc.so.6+0x9998a) (BuildId: f0b834daa3d05a80967e9ec2f990a1ea71c958fa)
    #5 0x7fb162f1d855 in gsignal (/lib/x86_64-linux-gnu/libc.so.6+0x42855) (BuildId: f0b834daa3d05a80967e9ec2f990a1ea71c958fa)
    #6 0x7fb162f018b6 in abort (/lib/x86_64-linux-gnu/libc.so.6+0x268b6) (BuildId: f0b834daa3d05a80967e9ec2f990a1ea71c958fa)
    #7 0x7fb162f017da  (/lib/x86_64-linux-gnu/libc.so.6+0x267da) (BuildId: f0b834daa3d05a80967e9ec2f990a1ea71c958fa)
    #8 0x7fb162f14185 in __assert_fail (/lib/x86_64-linux-gnu/libc.so.6+0x39185) (BuildId: f0b834daa3d05a80967e9ec2f990a1ea71c958fa)
    #9 0x55cd51accacc in autofile_fuzz_target(Span<unsigned char const>)::$_3::operator()() const src/test/fuzz/autofile.cpp:50:17
    #10 0x55cd51accacc in unsigned long CallOneOf<autofile_fuzz_target(Span<unsigned char const>)::$_0, autofile_fuzz_target(Span<unsigned char const>)::$_1, autofile_fuzz_target(Span<unsigned char const>)::$_2, autofile_fuzz_target(Span<unsigned char const>)::$_3, autofile_fuzz_target(Span<unsigned char const>)::$_4, autofile_fuzz_target(Span<unsigned char const>)::$_5>(FuzzedDataProvider&, autofile_fuzz_target(Span<unsigned char const>)::$_0, autofile_fuzz_target(Span<unsigned char const>)::$_1, autofile_fuzz_target(Span<unsigned char const>)::$_2, autofile_fuzz_target(Span<unsigned char const>)::$_3, autofile_fuzz_target(Span<unsigned char const>)::$_4, autofile_fuzz_target(Span<unsigned char const>)::$_5) src/./test/fuzz/util.h:42:27
    #11 0x55cd51accacc in autofile_fuzz_target(Span<unsigned char const>) src/test/fuzz/autofile.cpp:27:9
    #12 0x55cd5204a52c in std::function<void (Span<unsigned char const>)>::operator()(Span<unsigned char const>) const /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/bits/std_function.h:591:9
    #13 0x55cd5204a52c in LLVMFuzzerTestOneInput src/test/fuzz/fuzz.cpp:178:5
    #14 0x55cd51879814 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x199e814) (BuildId: 34c8e9fbac2706ae466ade7d359adebec5d82bcf)
    #15 0x55cd51878f09 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) (/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x199df09) (BuildId: 34c8e9fbac2706ae466ade7d359adebec5d82bcf)
    #16 0x55cd5187aaf6 in fuzzer::Fuzzer::ReadAndExecuteSeedCorpora(std::vector<fuzzer::SizedFile, std::allocator<fuzzer::SizedFile>>&) (/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x199faf6) (BuildId: 34c8e9fbac2706ae466ade7d359adebec5d82bcf)
    #17 0x55cd5187b017 in fuzzer::Fuzzer::Loop(std::vector<fuzzer::SizedFile, std::allocator<fuzzer::SizedFile>>&) (/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x19a0017) (BuildId: 34c8e9fbac2706ae466ade7d359adebec5d82bcf)
    #18 0x55cd5186875b in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x198d75b) (BuildId: 34c8e9fbac2706ae466ade7d359adebec5d82bcf)
    #19 0x55cd51892ce6 in main (/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x19b7ce6) (BuildId: 34c8e9fbac2706ae466ade7d359adebec5d82bcf)
    #20 0x7fb162f030cf  (/lib/x86_64-linux-gnu/libc.so.6+0x280cf) (BuildId: f0b834daa3d05a80967e9ec2f990a1ea71c958fa)
    #21 0x7fb162f03188 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x28188) (BuildId: f0b834daa3d05a80967e9ec2f990a1ea71c958fa)
    #22 0x55cd5185d624 in _start (/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x1982624) (BuildId: 34c8e9fbac2706ae466ade7d359adebec5d82bcf)
NOTE: libFuzzer has rudimentary signal handlers.
      Combine libFuzzer with AddressSanitizer or similar for better crash reports.
SUMMARY: libFuzzer: deadly signal
MS: 0 ; base unit: 0000000000000000000000000000000000000000
artifact_prefix='./'; Test unit written to ./crash-276f4f9d3e6ba98f897bd1082aec7214ca8e4f5d
INFO: Running with entropic power schedule (0xFF, 100).
INFO: Seed: 3765617163
INFO: Loaded 1 modules   (547059 inline 8-bit counters): 547059 [0x55cd54dd29b8, 0x55cd54e582ab), 
INFO: Loaded 1 PC tables (547059 PCs): 547059 [0x55cd54e582b0,0x55cd556b11e0), 
INFO:      535 files found in /ci_container_base/ci/scratch/qa-assets/fuzz_seed_corpus/autofile
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 1041296 bytes
INFO: seed corpus: files: 535 min: 1b max: 1041296b total: 17489928b rss: 106Mb
fuzz: test/fuzz/autofile.cpp:50: auto autofile_fuzz_target(FuzzBufferType)::(anonymous class)::operator()() const: Assertion `auto_file.fclose() == 0' failed.
==18378== ERROR: libFuzzer: deadly signal
    #0 0x55cd5193ab55 in __sanitizer_print_stack_trace (/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x1a5fb55) (BuildId: 34c8e9fbac2706ae466ade7d359adebec5d82bcf)
    #1 0x55cd518922fc in fuzzer::PrintStackTrace() (/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x19b72fc) (BuildId: 34c8e9fbac2706ae466ade7d359adebec5d82bcf)
    #2 0x55cd51878367 in fuzzer::Fuzzer::CrashCallback() (/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x199d367) (BuildId: 34c8e9fbac2706ae466ade7d359adebec5d82bcf)
    #3 0x7fb162f1d8ff  (/lib/x86_64-linux-gnu/libc.so.6+0x428ff) (BuildId: f0b834daa3d05a80967e9ec2f990a1ea71c958fa)
    #4 0x7fb162f7498a in pthread_kill (/lib/x86_64-linux-gnu/libc.so.6+0x9998a) (BuildId: f0b834daa3d05a80967e9ec2f990a1ea71c958fa)
    #5 0x7fb162f1d855 in gsignal (/lib/x86_64-linux-gnu/libc.so.6+0x42855) (BuildId: f0b834daa3d05a80967e9ec2f990a1ea71c958fa)
    #6 0x7fb162f018b6 in abort (/lib/x86_64-linux-gnu/libc.so.6+0x268b6) (BuildId: f0b834daa3d05a80967e9ec2f990a1ea71c958fa)
    #7 0x7fb162f017da  (/lib/x86_64-linux-gnu/libc.so.6+0x267da) (BuildId: f0b834daa3d05a80967e9ec2f990a1ea71c958fa)
    #8 0x7fb162f14185 in __assert_fail (/lib/x86_64-linux-gnu/libc.so.6+0x39185) (BuildId: f0b834daa3d05a80967e9ec2f990a1ea71c958fa)
    #9 0x55cd51accacc in autofile_fuzz_target(Span<unsigned char const>)::$_3::operator()() const src/test/fuzz/autofile.cpp:50:17
    #10 0x55cd51accacc in unsigned long CallOneOf<autofile_fuzz_target(Span<unsigned char const>)::$_0, autofile_fuzz_target(Span<unsigned char const>)::$_1, autofile_fuzz_target(Span<unsigned char const>)::$_2, autofile_fuzz_target(Span<unsigned char const>)::$_3, autofile_fuzz_target(Span<unsigned char const>)::$_4, autofile_fuzz_target(Span<unsigned char const>)::$_5>(FuzzedDataProvider&, autofile_fuzz_target(Span<unsigned char const>)::$_0, autofile_fuzz_target(Span<unsigned char const>)::$_1, autofile_fuzz_target(Span<unsigned char const>)::$_2, autofile_fuzz_target(Span<unsigned char const>)::$_3, autofile_fuzz_target(Span<unsigned char const>)::$_4, autofile_fuzz_target(Span<unsigned char const>)::$_5) src/./test/fuzz/util.h:42:27
    #11 0x55cd51accacc in autofile_fuzz_target(Span<unsigned char const>) src/test/fuzz/autofile.cpp:27:9
    #12 0x55cd5204a52c in std::function<void (Span<unsigned char const>)>::operator()(Span<unsigned char const>) const /usr/bin/../lib/gcc/x86_64-linux-gnu/13/../../../../include/c++/13/bits/std_function.h:591:9
    #13 0x55cd5204a52c in LLVMFuzzerTestOneInput src/test/fuzz/fuzz.cpp:178:5
    #14 0x55cd51879814 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) (/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x199e814) (BuildId: 34c8e9fbac2706ae466ade7d359adebec5d82bcf)
    #15 0x55cd51878f09 in fuzzer::Fuzzer::RunOne(unsigned char const*, unsigned long, bool, fuzzer::InputInfo*, bool, bool*) (/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x199df09) (BuildId: 34c8e9fbac2706ae466ade7d359adebec5d82bcf)
    #16 0x55cd5187aaf6 in fuzzer::Fuzzer::ReadAndExecuteSeedCorpora(std::vector<fuzzer::SizedFile, std::allocator<fuzzer::SizedFile>>&) (/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x199faf6) (BuildId: 34c8e9fbac2706ae466ade7d359adebec5d82bcf)
    #17 0x55cd5187b017 in fuzzer::Fuzzer::Loop(std::vector<fuzzer::SizedFile, std::allocator<fuzzer::SizedFile>>&) (/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x19a0017) (BuildId: 34c8e9fbac2706ae466ade7d359adebec5d82bcf)
    #18 0x55cd5186875b in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) (/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x198d75b) (BuildId: 34c8e9fbac2706ae466ade7d359adebec5d82bcf)
    #19 0x55cd51892ce6 in main (/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x19b7ce6) (BuildId: 34c8e9fbac2706ae466ade7d359adebec5d82bcf)
    #20 0x7fb162f030cf  (/lib/x86_64-linux-gnu/libc.so.6+0x280cf) (BuildId: f0b834daa3d05a80967e9ec2f990a1ea71c958fa)
    #21 0x7fb162f03188 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x28188) (BuildId: f0b834daa3d05a80967e9ec2f990a1ea71c958fa)
    #22 0x55cd5185d624 in _start (/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz+0x1982624) (BuildId: 34c8e9fbac2706ae466ade7d359adebec5d82bcf)
NOTE: libFuzzer has rudimentary signal handlers.
      Combine libFuzzer with AddressSanitizer or similar for better crash reports.
SUMMARY: libFuzzer: deadly signal
MS: 0 ; base unit: 0000000000000000000000000000000000000000
artifact_prefix='./'; Test unit written to ./crash-276f4f9d3e6ba98f897bd1082aec7214ca8e4f5d
Target ['/ci_container_base/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/test/fuzz/fuzz', '-runs=1', PosixPath('/ci_container_base/ci/scratch/qa-assets/fuzz_seed_corpus/autofile')] failed with exit code 77
```

No base64 of the unit in the output, I guess I cannot access `./crash-276f4f9d3e6ba98f897bd1082aec7214ca8e4f5d` and I cannot reproduce by running `FUZZ=autofile ./src/test/fuzz/fuzz -runs=1000000 .../qa-assets/fuzz_seed_corpus/autofile` locally. Maybe the CI machine ran out of disk space?

vasild replied:
I realized that `auto_file` used in this fuzz test is not associated with a regular file on the filesystem, but with a "cookie" returned by `fuzzed_file_provider.open()` returned by `fopencookie(3)` which uses a custom close function `FuzzedFileProvider::close()` which can arbitrarily return `-1`:

https://github.com/bitcoin/bitcoin/blob/207220ce8b767d8efdb5abf042ecf23d846ded73/src/test/fuzz/util.cpp#L352

changed this to ignore the return value of `auto_file.fclose()`.

vasild commented on src/streams.cpp:
```diff
@@ -4,14 +4,20 @@
-        return std::fread(dst.data(), 1, dst.size(), m_file);
+        const auto n = std::fread(dst.data(), 1, dst.size(), m_file);
+        if (ferror(m_file)) {
+            throw std::ios_base::failure(strprintf("AutoFile::detail_fread: %s", SysErrorString(errno)));
```
Alright, moved to the caller.

IMO this is a bugfix, but I am not 100% sure either. Might as well be viewed as "belt-and-suspenders". What is certain is that it can't hurt even if there is never an error on any platform if `detail_fread(dst) == dst.size()`. A unit test would require simulating an IO error under the `fread(3)` call and identically behaving `fread(3)` on all platforms.

vasild commented:
> I don't think any care close-check needs to be done when reading from a file.

That is my understanding too.
 
> So what about removing the `write` method from `AutoFile`, and introduce a new derived class to add it back. This class could `Assume` that the file was flushed/closed before the destructor is called?

`Assume` is more for code correctness, not for external errors (like IO error). If it does not fail during testing and on CI, that does not mean IO errors are absent and will not occur on the users' machines. Given that `Assume()` is removed from production builds, it still means that file corruption could remain unnoticed on live environments, like now on `master`.

vasild commented:
Then I misunderstood this:

> This class could Assume that the file was flushed/closed before the destructor is called

How?

vasild commented:
How/when would the file be closed?

vasild commented:
Ok, the users of the class then have to explicitly close. That is 1.4. from the OP, I elaborated it with more words.

achow101 commented:
Seems like there is some discussion in the description that is a better fit for a brainstorming issue.

vasild commented:
`5543990321...1fa61d9edc`: rebase and log a message from `~AutoFile()` if closing fails and terminate the program.

Sjors commented:
@vasild CI seems unhappy.

vasild commented:
Fixed the CI!

vasild commented on src/streams.cpp:
```diff
@@ -23,8 +25,9 @@ std::size_t AutoFile::detail_fread(Span<std::byte> dst)
 {
-    if (detail_fread(dst) != dst.size()) {
-        throw std::ios_base::failure(feof() ? "AutoFile::read: end of file" : "AutoFile::read: fread failed");
+    if (detail_fread(dst) != dst.size() || ferror(m_file)) {
```
This was somewhat answered in https://github.com/bitcoin/bitcoin/pull/29307#discussion_r1465152180 above but let me elaborate - the thing that prompted me to do this change is this text from the `fread(3)` manual page:

> The function fread() does not distinguish between end-of-file and error,
> and callers must use feof(3) and ferror(3) to determine which occurred.

I think checking for `ferror(3)` after `fread(3)` is a kind of harmless "belt-and-suspenders". If you are sure that if `detail_fread(dst) == dst.size()` then `ferror(m_file)` will always be `false` or feel strongly against this then I could drop the commit de23848eed `util: check for error after reading from a file`.

vasild replied:
Well, it is pretty much obvious that checking for an error after read is either:

* a noop, if the read could have never failed because it managed to read as many bytes as requested or
* a bug fix where a read error was ignored before the fix

so this is a safe change (it does not break anything), possibly fixing a bug. Given that you keep pushing against this I will drop this commit in the hopes to move this PR forward.

vasild replied:
I just found a similar code to this proposed change:

https://github.com/bitcoin/bitcoin/blob/2d21060af831fa8f8f1791b4464961d5f28a88cb/src/node/utxo_snapshot.cpp#L74-L80

After reading from a file and getting everything that is needed from it, the code checks if there is an extra data and if that read fails it reports an error.

vasild replied:
I think I see your point, you prefer 1. over 2., assuming we got all the bytes we asked for from a read, then:

1. Ignore possible IO error by not checking for it
2. Check for an error and behave as if the read failed

I dropped this change.

vasild commented on src/kernel/mempool_persist.cpp:
```diff
@@ -213,6 +218,7 @@ bool DumpMempool(const CTxMemPool& pool, const fs::path& dump_path, FopenFn mock
                   fs::file_size(dump_path));
     } catch (const std::exception& e) {
         LogInfo("Failed to dump mempool: %s. Continuing anyway.\n", e.what());
+        (void)file.fclose();
```
It is best if all users of `AutoFile` explicitly call the `fclose()` method and check that it succeeds (option 3. from PR description). This is not feasible because there are many users of `AutoFile`. However, some of those users already do call the `fclose()` method explicitly. For those users it is easy and I added checks whether `fclose()` succeeds. `DumpMempool()` is such a user and the change is:

```diff
             throw std::runtime_error("FileCommit failed");
-        file.fclose();           
+        if (file.fclose() != 0) {    
+            throw std::runtime_error(
+                strprintf("Error closing %s: %s", fs::PathToString(file_fspath), SysErrorString(errno)));
+        }
         if (!RenameOver(dump_path + ".new", dump_path)) {
```

For consistency in `DumpMempool()` I changed all code paths in that function to explicitly call `fclose()`. There are two codepaths and this added `(void)file.fclose();` is the other one. I ignore the return value because this is already failure handling and the function is going to return `false`. I didn't want to do something like "while handling error1, error2 occurred".

vasild replied:
Resolving this discussion. Changed the semantic of `AutoFile` to require all write users to call `fclose()`.

vasild commented on src/streams.h:
```diff
@@ -383,26 +385,42 @@ class BitStreamWriter
+    {
+        if (fclose() != 0 && m_was_written) {
+            LogPrintLevel(BCLog::ALL, BCLog::Level::Error, "Error closing file: %s\n", SysErrorString(errno));
+            std::abort();
```
> if (m_was_written) Assume(IsNull());

Done.

maflcko reviewed (commented):
lgtm ACK 11be9f4103ea219f801a1f0fe1385f66ca70ad22

Will do some testing later

vasild commented:
`11be9f4103...ea89a6e368`: rebase due to conflicts

vasild commented:
`ea89a6e368...4533445fd7`: rebase due to conflicts

vasild commented:
CI failure seems unrelated.

ryanofsky commented:
The PR description is a bit vague about how it is changing the AutoFile API.

Previously, the autofile API *allowed* but did not *require* AutoFile callers to check for failing `fclose` calls after writing. Now it requires that callers call the `fclose` method manually and explicitly use or discard its return value.

Checking for failing `fclose` calls is important because if a call fails, it probably indicates that there is a serious problem that the should trigger a log message or an alert. Specifically if the file was used for writing it could mean that data could not successfully be saved to disk. So it would be bad for `fclose` errors to be discarded without being handled at all or reported somewhere.

The PR makes AutoFile require callers to handle failing fclose calls, and it enforces this in two ways:

1. It adds `[[nodiscard]]` annotation to the `fclose` method so its return value cannot be ignored. The return value is 0 if successful or `EOF` (-1) if unsuccessful.
2. It tries to enforce that callers actually invoke `fclose` by adding an `Assume` statement to the destructor that terminates the program with a fatal error if the file was written to and fclose was not called and the binary was built in a debug mode.

It seems like this approach works, but it is maybe a little fragile because it doesn't actually enforce that caller handles the fclose result in every code path. So for example if an early return statement is added after a caller opens an autofile object but before it calls the fclose method, and the early return code path is not normally executed, it can introduce a latent bug that occasionally will crash debug builds, while providing no indication that anything is wrong in release builds.

I think if the goal is to require AutoFile callers to handle failing fclose calls, a more reliable way to do that would be to change the constructor signature to

```c++
explicit AutoFile(std::FILE* file, std::function<void()> on_error);
```

And require the caller to provide an on_error callback that does error handing like

```c++
[&] {
        remove(pathTmp);
        LogError("%s: Failed to close file %s: %s\n", __func__, fs::PathToString(pathTmp), SysErrorString(errno));
}
```

I think the current approach is ok too, but it seems more fragile because it doesn't detect all programming errors, and when it does it detects them with a runtime crash instead of a compile time error.

vasild commented:
@ryanofsky, I am not so attached to the current approach :) if it can be improved, then lets improve it.

For the `on_error` callback - that would be called on fclose() error from destructor only, right? Not on any error? E.g. the `write()` method throws an exception on error.

Not all callers have the file name when creating `AutoFile` which means we would have to log some generic message without a file name. Or go an extra mile and make sure we have the file name, but some callers don't use _files_ at all, IIRC the fuzzer uses some cookie "files" that are not on the filesystem.

Then the callback can be omitted and the message can be logged from the destructor:

```diff
     ~AutoFile()
     {   
         if (m_was_written) {
             // Callers that wrote to the file must have closed it explicitly
             // with the fclose() method and checked that the close succeeded.
             // This is because here from the destructor we have no way to signal
             // error due to close which, after write, could mean the file is
             // corrupted and must be handled properly at the call site.
             Assume(IsNull());
         }
                        
-        (void)fclose();
+        if (fclose() != 0) {
+            LogError("Failed to close file: %s\n", SysErrorString(errno));
+        }
     }
```

vasild commented:
`4533445fd7...1dcd626fe1`: rebase due to conflicts

vasild commented:
`1dcd626fe1...587d996f07`: rebase, get the extra printout from https://github.com/bitcoin/bitcoin/pull/29307#issuecomment-2245216013 and resolve a silent merge conflict with https://github.com/bitcoin/bitcoin/pull/28052

vasild commented:
`587d996f07...489c2477e7`: rebase to pick CMake

vasild commented:
`489c2477e7...dba7835386`: rebase due to conflicts

hodlinator reviewed (commented):
Concept ACK

Appreciate the intent of being rigorous in reporting issues with flushing file content to disk.

#### Alternative implementation

The current PR which produces a *run-time failure* in the `AutoFile` destructor upon forgetting to explicitly call `fclose()`. I attempted an alternative implementation ([branch](https://github.com/bitcoin/bitcoin/compare/master...hodlinator:bitcoin:pr/29307_suggestions_prime)) [inspired by ryanofsky's comment](https://github.com/bitcoin/bitcoin/pull/29307#issuecomment-2243842400) that takes a lambda in the constructor which is called upon `fclose()`-failure in the destructor. This enforces the context-aware handling of the issue at *compile-time*. The change was somewhat invasive but I'm curious to hear what others make of it. (It also splits `AutoFile` into `FileReader` and `FileWriter` (naming inspired by other types in *streams.h*)).

vasild commented on src/rpc/blockchain.cpp:
```diff
@@ -47,6 +47,7 @@
+#include <util/syserror.h>
 #include <util/translation.h>
 #include <validation.h>
 #include <validationinterface.h>
```
It is being closed inside `WriteUTXOSnapshot()`.

vasild commented:
`dba7835...9c16900`: rebase due to conflicts

vasild commented:
`9c16900b9b...8939947449`: address suggestions

vasild commented on src/addrdb.cpp:
```diff
@@ -61,25 +62,28 @@ bool SerializeFileDB(const std::string& prefix, const fs::path& path, const Data
-    fileout.fclose();
+    if (fileout.fclose() != 0) {
+        remove(pathTmp);
+        LogError("%s: Failed to close file %s: %s\n", __func__, fs::PathToString(pathTmp), SysErrorString(errno));
```
Done. Note however that `fclose()` returns either `0` or `EOF` (`-1`). It does not return the error code.

vasild commented on src/index/blockfilterindex.cpp:
```diff
@@ -229,6 +229,12 @@ size_t BlockFilterIndex::WriteFilterToDisk(FlatFilePos& pos, const BlockFilter&
     fileout << filter.GetBlockHash() << filter.GetEncodedFilter();
+
+    if (fileout.fclose() != 0) {
+        LogPrintf("%s: Failed to close filter file %d\n", __func__, pos.nFile);
```
Done, modulo `fclose()` does not return the error code which is to be found in `errno`.

vasild commented on src/policy/fees.cpp:
```diff
@@ -950,7 +950,7 @@ void CBlockPolicyEstimator::Flush() {
     AutoFile est_file{fsbridge::fopen(m_estimation_filepath, "wb")};
-    if (est_file.IsNull() || !Write(est_file)) {
+    if (est_file.IsNull() || !Write(est_file) || est_file.fclose() != 0) {
         LogPrintf("Failed to write fee estimates to %s. Continue anyway.\n", fs::PathToString(m_estimation_filepath));
```
Done.

vasild commented:
> lambda in the constructor which is called upon fclose()-failure in the destructor ... I'm curious to hear what others make of it. (It also splits AutoFile into FileReader and FileWriter (naming inspired by other types in streams.h)).

Concept ACK, but I am worried that the lack of review interest here would mean that an even more invasive patch would never make it.

l0rinc commented:
Concept ACK

vasild commented:
`8939947449...aa88c6dfb6`: rebase due to conflicts

vasild commented:
`aa88c6dfb6...0e97a68e13`: fix CI

vasild commented:
`0e97a68e13...1f71111e21`: fix AI and CI (even more)

vasild commented:
`1f71111e21...c52b673bf8`: https://github.com/bitcoin/bitcoin/pull/29307#discussion_r2048480671

l0rinc reviewed (changes requested):
Concept ACK, we definitely want to know if this happens!

* It seems we're using the return value as a boolean - can make make it return one?
* The error messages seem all over the place, can we unify them to `LogError`?
* Not sure if scope-ing the AutoFile + closing it explicitly is the best solution -  would probably prefer @ryanofsky's `std::function<void()> on_error` or @maflcko's recommendation to let `AutoFile` guarantee proper closing.
* Sometimes we're explicitly showing a closing related failure, other times we're reusing previous error messages - it's not a very important or likely scenario, we could likely add it to other erroneous cases.
* We have discernible changes here, all in a single commit - would prefer separating the concerns a bit more with better commit message explanations

vasild committed:
util: check that a file has been closed before ~AutoFile() is called

If an `AutoFile` has been written to, then expect callers to have closed
it explicitly via the `AutoFile::fclose()` method. This is because if
the destructor calls `std::fclose()` and encounters an error, then it
is too late to indicate this to the caller in a meaningful way.

vasild commented:
`c52b673bf8...c9cd7d3aa9`: rebase and address suggestions

vasild commented on src/index/blockfilterindex.cpp:
```diff
@@ -151,7 +151,7 @@ bool BlockFilterIndex::CustomCommit(CDBBatch& batch)
         return false;
     }
-    if (!file.Commit()) {
+    if (!file.Commit() || file.fclose() != 0) {
```
There is even a bigger problem with `if (!file.Commit() || file.fclose() != 0) {` -- if the commit fails when `fclose()` will not be called, it will return and the file destructor will be upset that the file has been written to and is not closed. Fixed.

vasild commented on src/index/blockfilterindex.cpp:
```diff
@@ -229,6 +229,12 @@ size_t BlockFilterIndex::WriteFilterToDisk(FlatFilePos& pos, const BlockFilter&
+    if (fileout.fclose() != 0) {
+        LogPrintf("Failed to close filter file %d: %s", pos.nFile, SysErrorString(errno));
+        return 0;
+    }
```
Changed to `LogError()`, but `__func__` is redundant with the config option `-logsourcelocations`, so I omitted it.

vasild commented on src/node/blockstorage.cpp:
```diff
@@ -947,20 +947,30 @@ bool BlockManager::WriteBlockUndo(const CBlockUndo& blockundo, BlockValidationSt
             }
 
-            fileout.flush(); // Make sure `AutoFile`/`BufferedWriter` go out of scope before we call `FlushUndoFile`
+            if (file.fclose() != 0) {
```
Right. Removed the scope.

vasild commented on src/node/blockstorage.cpp:
```diff
@@ -947,20 +947,30 @@ bool BlockManager::WriteBlockUndo(const CBlockUndo& blockundo, BlockValidationSt
+                              SysErrorString(errno));
+                return FatalError(m_opts.notifications, state, _("Failed to close block undo file."));
+            }
+            // Make sure that the file is closed before we call `FlushUndoFile`.
```
Moved the comment before the actual closing. Reduced as well.

vasild commented on src/node/blockstorage.cpp:
```diff
@@ -1142,6 +1165,11 @@ static auto InitBlocksdirXorKey(const BlockManager::Options& opts)
         )};
         xor_key_file << xor_key;
+        if (xor_key_file.fclose() != 0) {
+            throw std::runtime_error{strprintf("Error closing XOR key file %s: %s\n",
```
Done.

vasild commented on src/test/fuzz/autofile.cpp:
```diff
@@ -62,5 +62,10 @@ FUZZ_TARGET(autofile)
+    } else {
+        // FuzzedFileProvider::close() is expected to fail sometimes. Don't let
+        // the destructor of AutoFile be upset by a failing fclose(). Close it
+        // explicitly (and ignore any errors) so that the destructor is a noop.
```
Removed this outdated comment.

vasild commented on src/streams.h:
```diff
@@ -386,27 +389,48 @@ class BitStreamWriter
     std::FILE* m_file;
     std::vector<std::byte> m_xor;
     std::optional<int64_t> m_position;
+    bool m_was_written{false};
```
Done.

vasild commented on src/streams.h:
```diff
@@ -386,27 +389,48 @@ class BitStreamWriter
+        }
+
+        if (fclose() != 0) {
+            LogPrintLevel(BCLog::ALL, BCLog::Level::Error, "Failed to close file: %s", SysErrorString(errno));
```
Done.

vasild commented on src/streams.h:
```diff
@@ -386,27 +389,48 @@ class BitStreamWriter
+        if (m_was_written) {
+            // Callers that wrote to the file must have closed it explicitly
+            // with the fclose() method and checked that the close succeeded.
+            // This is because here from the destructor we have no way to signal
+            // error due to close which, after write, could mean the file is
```
Elaborated a little bit:

```cpp
            // Callers that wrote to the file must have closed it explicitly
            // with the fclose() method and checked that the close succeeded.
            // This is because here from the destructor we have no way to signal
            // error due to close which, after write, could mean the file is
            // corrupted and must be handled properly at the call site.
            // Destructors in C++ cannot signal an error to the callers because
            // they do not return a value and are not allowed to throw exceptions.
```

l0rinc replied:
> because here from the destructor

Still not sure what this means exactly, would this maybe work better?

```suggestion
            // This is because here in the destructor we have no way to signal
            // errors from close(), which, after write, could mean the file is
```

vasild replied:
Done

vasild commented on src/test/flatfile_tests.cpp:
```diff
@@ -79,13 +81,15 @@ BOOST_AUTO_TEST_CASE(flatfile_open)
 
         file >> LIMITED_STRING(text, 256);
         BOOST_CHECK_EQUAL(text, line2);
+        BOOST_REQUIRE_EQUAL(file.fclose(), 0);
```
I think it is better to stop the test immediately if closing a file gives an error.

vasild commented on src/streams.cpp:
```diff
@@ -95,6 +95,7 @@ void AutoFile::write(std::span<const std::byte> src)
             src = src.subspan(buf_now.size());
         }
     }
+    m_was_written = true;
```
What to be added to `AutoFile::write_buffer`?

l0rinc replied:
`m_was_written = true`, since `AutoFile::write_buffer` also dirties (which is the term we use in other cases, if you think it describes the behavior better, consider renaming)

vasild replied:
I see. Done. I also moved the `m_was_written = true` to be just after the `fwrite()` calls.

vasild commented:
`66087035cf...fe2d1955f6`: add missing includes in the first commit

l0rinc commented on src/streams.h:
```diff
@@ -386,27 +389,48 @@ class BitStreamWriter
     bool feof() const { return std::feof(m_file); }
 
-    int fclose()
+    [[nodiscard]] int fclose()
```
I find returning error codes for boolean values a lot more confusing.

hodlinator replied:
In that case it would be better to call it `bool AutoFile::Close()` rather than `AutoFile::fclose`.

l0rinc replied:
Agree, it's what I wrote in the review as well:
>  the current `fclose` returns a fake error code, but is actually a boolean - may make sense to rename it (e.g. `Close`, similarly to how we've wrapped `ftruncate` with `TruncateFile`) and migrate to returning a boolean to avoid the current confusion.

l0rinc reviewed (changes requested):
Code review fe2d1955f6333c57f54aed2a2a8b3cdaa6bf9d71

* blocker: the new `AutoFile::write_buffer` should also mark the file as dirty.
* the current `fclose` returns a fake error code, but is actually a boolean - may make sense to rename it (e.g. `Close`, similarly to how we've wrapped `ftruncate` with `TruncateFile`) and migrate to returning a boolean to avoid the current confusion.

<details>
<summary>Diff since last review</summary>

```patch
diff --git a/src/index/blockfilterindex.cpp b/src/index/blockfilterindex.cpp
index 9e6ea8d7d0..6b47a0fa0a 100644
--- a/src/index/blockfilterindex.cpp
+++ b/src/index/blockfilterindex.cpp
@@ -13,6 +13,7 @@
 #include <node/blockstorage.h>
 #include <undo.h>
 #include <util/fs_helpers.h>
+#include <util/syserror.h>
 #include <validation.h>
 
 /* The index database stores three items for each block: the disk location of the encoded filter,
@@ -151,8 +152,13 @@ bool BlockFilterIndex::CustomCommit(CDBBatch& batch)
         LogError("%s: Failed to open filter file %d\n", __func__, pos.nFile);
         return false;
     }
-    if (!file.Commit() || file.fclose() != 0) {
+    if (!file.Commit()) {
         LogError("%s: Failed to commit filter file %d\n", __func__, pos.nFile);
+        (void)file.fclose();
+        return false;
+    }
+    if (file.fclose() != 0) {
+        LogError("Failed to close filter file %d after commit: %s", pos.nFile, SysErrorString(errno));
         return false;
     }
 
@@ -205,8 +211,13 @@ size_t BlockFilterIndex::WriteFilterToDisk(FlatFilePos& pos, const BlockFilter&
             LogPrintf("%s: Failed to truncate filter file %d\n", __func__, pos.nFile);
             return 0;
         }
-        if (!last_file.Commit() || last_file.fclose() != 0) {
+        if (!last_file.Commit()) {
             LogPrintf("%s: Failed to commit filter file %d\n", __func__, pos.nFile);
+            (void)last_file.fclose();
+            return 0;
+        }
+        if (last_file.fclose() != 0) {
+            LogError("Failed to close filter file %d after commit: %s", pos.nFile, SysErrorString(errno));
             return 0;
         }
 
@@ -231,7 +242,7 @@ size_t BlockFilterIndex::WriteFilterToDisk(FlatFilePos& pos, const BlockFilter&
     fileout << filter.GetBlockHash() << filter.GetEncodedFilter();
 
     if (fileout.fclose() != 0) {
-        LogPrintf("Failed to close filter file %d: %s", pos.nFile, SysErrorString(errno));
+        LogError("Failed to close filter file %d: %s", pos.nFile, SysErrorString(errno));
         return 0;
     }
 
diff --git a/src/node/blockstorage.cpp b/src/node/blockstorage.cpp
index 8b21356c42..e6a979b5f3 100644
--- a/src/node/blockstorage.cpp
+++ b/src/node/blockstorage.cpp
@@ -33,6 +33,7 @@
 #include <util/fs.h>
 #include <util/signalinterrupt.h>
 #include <util/strencodings.h>
+#include <util/syserror.h>
 #include <util/translation.h>
 #include <validation.h>
 
@@ -940,37 +941,31 @@ bool BlockManager::WriteBlockUndo(const CBlockUndo& blockundo, BlockValidationSt
             return false;
         }
 
+        // Open history file to append
+        AutoFile file{OpenUndoFile(pos)};
+        if (file.IsNull()) {
+            LogError("OpenUndoFile failed for %s while writing block undo", pos.ToString());
+            return FatalError(m_opts.notifications, state, _("Failed to write undo data."));
+        }
         {
-            // Open history file to append
-            AutoFile file{OpenUndoFile(pos)};
-            if (file.IsNull()) {
-                LogError("OpenUndoFile failed for %s while writing block undo", pos.ToString());
-                return FatalError(m_opts.notifications, state, _("Failed to write undo data."));
-            }
+            BufferedWriter fileout{file};
+
+            // Write index header
+            fileout << GetParams().MessageStart() << blockundo_size;
+            pos.nPos += STORAGE_HEADER_BYTES;
             {
-                BufferedWriter fileout{file};
-
-                // Write index header
-                fileout << GetParams().MessageStart() << blockundo_size;
-                pos.nPos += STORAGE_HEADER_BYTES;
-                {
-                    // Calculate checksum
-                    HashWriter hasher{};
-                    hasher << block.pprev->GetBlockHash() << blockundo;
-                    // Write undo data & checksum
-                    fileout << blockundo << hasher.GetHash();
-                }
+                // Calculate checksum
+                HashWriter hasher{};
+                hasher << block.pprev->GetBlockHash() << blockundo;
+                // Write undo data & checksum
+                fileout << blockundo << hasher.GetHash();
             }
+        }
 
-            if (file.fclose() != 0) {
-                LogPrintLevel(BCLog::BLOCKSTORAGE,
-                              BCLog::Level::Error,
-                              "Failed to close block undo file %s: %s",
-                              pos.ToString(),
-                              SysErrorString(errno));
-                return FatalError(m_opts.notifications, state, _("Failed to close block undo file."));
-            }
-            // Make sure that the file is closed before we call `FlushUndoFile`.
+        // Make sure that the file is closed before we call `FlushUndoFile`.
+        if (file.fclose() != 0) {
+            LogError("Failed to close block undo file %s: %s", pos.ToString(), SysErrorString(errno));
+            return FatalError(m_opts.notifications, state, _("Failed to close block undo file."));
         }
 
         // rev files are written in block height order, whereas blk files are written as blocks come in (often out of order)
@@ -1114,11 +1109,7 @@ FlatFilePos BlockManager::WriteBlock(const CBlock& block, int nHeight)
     }
 
     if (file.fclose() != 0) {
-        LogPrintLevel(BCLog::BLOCKSTORAGE,
-                      BCLog::Level::Error,
-                      "Failed to close block file %s: %s",
-                      pos.ToString(),
-                      SysErrorString(errno));
+        LogError("Failed to close block file %s: %s", pos.ToString(), SysErrorString(errno));
         m_opts.notifications.fatalError(_("Failed to close file when writing block."));
         return FlatFilePos();
     }
@@ -1166,7 +1157,7 @@ static auto InitBlocksdirXorKey(const BlockManager::Options& opts)
         )};
         xor_key_file << xor_key;
         if (xor_key_file.fclose() != 0) {
-            throw std::runtime_error{strprintf("Error closing XOR key file %s: %s\n",
+            throw std::runtime_error{strprintf("Error closing XOR key file %s: %s",
                                                fs::PathToString(xor_key_path),
                                                SysErrorString(errno))};
         }
diff --git a/src/policy/fees.cpp b/src/policy/fees.cpp
index ea763efc98..3a59930c4b 100644
--- a/src/policy/fees.cpp
+++ b/src/policy/fees.cpp
@@ -19,6 +19,7 @@
 #include <uint256.h>
 #include <util/fs.h>
 #include <util/serfloat.h>
+#include <util/syserror.h>
 #include <util/time.h>
 
 #include <algorithm>
@@ -953,11 +954,16 @@ void CBlockPolicyEstimator::Flush() {
 void CBlockPolicyEstimator::FlushFeeEstimates()
 {
     AutoFile est_file{fsbridge::fopen(m_estimation_filepath, "wb")};
-    if (est_file.IsNull() || !Write(est_file) || est_file.fclose() != 0) {
-        LogWarning("Failed to write fee estimates to %s. Continuing anyway.", fs::PathToString(m_estimation_filepath));
-    } else {
-        LogPrintf("Flushed fee estimates to %s.\n", fs::PathToString(m_estimation_filepath.filename()));
+    if (est_file.IsNull() || !Write(est_file)) {
+        LogPrintf("Failed to write fee estimates to %s. Continue anyway.\n", fs::PathToString(m_estimation_filepath));
+        (void)est_file.fclose();
+        return;
+    }
+    if (est_file.fclose() != 0) {
+        LogError("Failed to close fee estimates file %s: %s. Continuing anyway.", fs::PathToString(m_estimation_filepath), SysErrorString(errno));
+        return;
     }
+    LogPrintf("Flushed fee estimates to %s.\n", fs::PathToString(m_estimation_filepath.filename()));
 }
 
 bool CBlockPolicyEstimator::Write(AutoFile& fileout) const
diff --git a/src/streams.h b/src/streams.h
index c13cebc5c6..09b2e20c08 100644
--- a/src/streams.h
+++ b/src/streams.h
@@ -416,11 +416,13 @@ public:
             // This is because here from the destructor we have no way to signal
             // error due to close which, after write, could mean the file is
             // corrupted and must be handled properly at the call site.
+            // Destructors in C++ cannot signal an error to the callers because
+            // they do not return a value and are not allowed to throw exceptions.
             Assume(IsNull());
         }
 
         if (fclose() != 0) {
-            LogPrintLevel(BCLog::ALL, BCLog::Level::Error, "Failed to close file: %s", SysErrorString(errno));
+            LogError("Failed to close file: %s", SysErrorString(errno));
         }
     }
 
diff --git a/src/test/fuzz/autofile.cpp b/src/test/fuzz/autofile.cpp
index cdbae9c3cc..8d17624da6 100644
--- a/src/test/fuzz/autofile.cpp
+++ b/src/test/fuzz/autofile.cpp
@@ -63,9 +63,6 @@ FUZZ_TARGET(autofile)
             fclose(f);
         }
     } else {
-        // FuzzedFileProvider::close() is expected to fail sometimes. Don't let
-        // the destructor of AutoFile be upset by a failing fclose(). Close it
-        // explicitly (and ignore any errors) so that the destructor is a noop.
         (void)auto_file.fclose();
     }
 }
```

</details>

l0rinc commented on src/node/blockstorage.cpp:
```diff
@@ -959,8 +960,12 @@ bool BlockManager::WriteBlockUndo(const CBlockUndo& blockundo, BlockValidationSt
-            fileout.flush(); // Make sure `AutoFile`/`BufferedWriter` go out of scope before we call `FlushUndoFile`
+        // Make sure that the file is closed before we call `FlushUndoFile`.
+        if (file.fclose() != 0) {
+            LogError("Failed to close block undo file %s: %s", pos.ToString(), SysErrorString(errno));
+            return FatalError(m_opts.notifications, state, _("Failed to close block undo file."));
```
do we need both messages? How will this be displayed in the logs, isn't it redundant?

vasild replied:
This just mimics the error handling when the file cannot be opened:

https://github.com/bitcoin/bitcoin/blob/53ccb75f0c47e1e48504cb392d7589548a8d25a2/src/node/blockstorage.cpp#L945-L949

vasild commented:
`fe2d1955f6...2c1a8ff0c0`: address suggestions

> AutoFile::write_buffer should also mark the file as dirty

Done.

> the current fclose returns a fake error code...

Replied to the discussion thread at https://github.com/bitcoin/bitcoin/pull/29307#discussion_r2048792563

vasild commented:
`2c1a8ff0c0...61ba3fc494`: fix CI

l0rinc commented:
`write_buffer` was also marked as dirty in latest push.

untested ACK 61ba3fc494f2f5e60b668316c84cecbdc123e5fa

Q: shouldn't `Truncate` and `Commit` also set `m_was_written = true`?

vasild commented:
`61ba3fc494...9057ab8b80`:

> shouldn't `Truncate` and `Commit` also set `m_was_written = true`?

Good question! I think `Truncate()` - yes because it changes the contents of the file. Changed.

`Commit()` does not, so I left it alone. In other words - if a file is opened, none of the `write()`, `write_buffer()` or `Truncate()` is called, `Commit()` is called, and the file object is destroyed, could that be a problem? I think no because the contents of the file was not changed.

hodlinator reviewed (commented):
Code review 9057ab8b80137a89df68a3ed4bc49f9926e0f634

I like how the `AutoFile` change now is broken out into a 2nd commit.

hodlinator commented on src/test/fuzz/autofile.cpp:
```diff
@@ -47,7 +47,7 @@ FUZZ_TARGET(autofile)
             },
             [&] {
-                auto_file.fclose();
+                (void)auto_file.fclose();
```
Not really, seems like it was changed in that specific file and resolved without being changed more broadly.

vasild commented on src/bench/streams_findbyte.cpp:
```diff
@@ -28,7 +28,7 @@ static void FindByte(benchmark::Bench& bench)
 
     // Cleanup
-    file.fclose();
+    (void)file.fclose();
```
Note that this here is in a bench test where we operate on a real file and an error to close it means something really went wrong badly, e.g. a disk IO error. Such a condition should stop the test with a failure.

In contrast, in fuzz tests some of the "files" we operate on are fuzzed files which mean they are not really backed on disk and their operations are mocked to sometimes return an error, e.g. to simulate a broken disk. Those mocked errors are to be expected and should not make the test fail.

Those fuzz mocks are done via `FuzzedFileProvider::close()` which is set by `FuzzedFileProvider::open()` as a close callback of the mocked `FILE*` that is returned.

hodlinator committed:
rpc: take ownership of the file by WriteUTXOSnapshot()

Have `WriteUTXOSnapshot()` take rvalue reference to make it obvious that
it takes ownership of the file.

vasild committed:
flatfile: check whether the file has been closed successfully

vasild commented:
`9057ab8b80...c7b68dceb2`: address suggestions

vasild commented on src/node/blockstorage.cpp:
```diff
@@ -959,8 +960,12 @@ bool BlockManager::WriteBlockUndo(const CBlockUndo& blockundo, BlockValidationSt
                 // Write undo data & checksum
                 fileout << blockundo << hasher.GetHash();
             }
+        }
```
Done.

vasild commented on src/node/blockstorage.cpp:
```diff
@@ -959,8 +960,12 @@ bool BlockManager::WriteBlockUndo(const CBlockUndo& blockundo, BlockValidationSt
+            return FatalError(m_opts.notifications, state, _("Failed to close block undo file."));
         }
 
         // rev files are written in block height order, whereas blk files are written as blocks come in (often out of order)
```
Yes. Added one more commit to handle those.

vasild commented on src/node/mempool_persist.cpp:
```diff
@@ -201,7 +203,10 @@ bool DumpMempool(const CTxMemPool& pool, const fs::path& dump_path, FopenFn mock
 
         if (!skip_file_commit && !file.Commit())
             throw std::runtime_error("Commit failed");
```
> This throw would trigger the Assume() in ~AutoFile() to fail, no?

Yes. Added a `fclose()` call before the `throw` to address that.

I will consider the lambda in the destructor again. From the `fclose()` callers, I counted:

* 7 that ignore the `fclose()` return value
* 7 that log message + take a different code path
* 4 throw

Taking a different code path upon `fclose()` failure would be tricky when that is called from the `AutoFile` destructor. It would require code like:

```cpp
AutoFile a;
... possibly a lot of code ...
if (a.fclose() != 0) {
  LogError();
  return false;
}
return true;
```
to be written like:
```cpp
bool fclose_failed{false};
{
    AutoFile a{..., [&fclose_failed] () { fclose_failed = true; }};
    ... possibly a lot of code, all of it now indented further ...
}
if (fclose_failed) {
    LogError();
    return false;
}
return true;
```
I like the idea that the caller cannot easily forget to handle a `fclose()` failure though.

hodlinator reviewed (commented):
Code review c7b68dceb29ef67796f7edf79a161bbda766169c

This is a clear incremental improvement over what is on master.

Thanks for taking my suggestions!

Don't foresee any further issues preventing an A-C-K beyond the inline comments below.

vasild commented:
`c7b68dceb2...004ec6968e`: rebase and address suggestion

vasild commented:
`004ec6968e...f8a7580a08`: rebase due to conflicts

hodlinator reviewed (approved):
ACK f8a7580a0816586f75ba9b0e79300c893d362f9f

Mainly changes `AutoFile` to encourage judicious error checking upon file closure, surfacing failures to complete writing files to disk.

Several alternatives have been considered (listed in PR description) and I agree this is probably the best option for now.

l0rinc commented on src/addrdb.cpp:
```diff
@@ -61,25 +62,29 @@ bool SerializeFileDB(const std::string& prefix, const fs::path& path, const Data
+    if (fileout.fclose() != 0) {
+        const int errno_save{errno};
+        remove(pathTmp);
+        LogError("Failed to close file %s: %s", fs::PathToString(pathTmp), SysErrorString(errno_save));
```
In other cased you've added `after commit` to clarify when closing failed.

Also previous errors we've prefixed with `__func__`:
```suggestion
        LogError("s: Failed to close file %s after commit: %s", __func__, fs::PathToString(pathTmp), SysErrorString(errno_save));
```

Looking at the other examples you've changed, I assume this was deliberate - I'd still prefer unifying them, but I don't mind tackling it in a separate PR either.

l0rinc replied:
> __func__ is redundant

Yes, would be great to unify the affected messages - either by adding or removing in every touched method. But we can of course do it in a separate PR as well.

l0rinc commented on src/test/streams_tests.cpp:
```diff
@@ -379,8 +380,8 @@ BOOST_AUTO_TEST_CASE(streams_buffered_file)
 
-    // We can explicitly close the file, or the destructor will do it.
-    file.fclose();
+    // Explicitly close the file and check that the close succeeds.
```
comments isn't useful anymore, the code already states the same

vasild replied:
Removed.

l0rinc reviewed (approved):
ACK f8a7580a0816586f75ba9b0e79300c893d362f9f

Changes since last review:
* adjust `AutoFile` parameter move semantics;
* log unsuccessful closes as a last attempt;
* add `m_was_written` to `Truncate`;
* & explanatory comments.

<details>
<summary>Details</summary>

```patch
diff --git a/src/flatfile.cpp b/src/flatfile.cpp
index 388b30efae..df6596e940 100644
--- a/src/flatfile.cpp
+++ b/src/flatfile.cpp
@@ -46,7 +46,9 @@ FILE* FlatFileSeq::Open(const FlatFilePos& pos, bool read_only) const
     }
     if (pos.nPos && fseek(file, pos.nPos, SEEK_SET)) {
         LogPrintf("Unable to seek to position %u of %s\n", pos.nPos, fs::PathToString(path));
-        fclose(file);
+        if (fclose(file) != 0) {
+            LogError("Unable to close file %s", fs::PathToString(path));
+        }
         return nullptr;
     }
     return file;
@@ -68,7 +70,10 @@ size_t FlatFileSeq::Allocate(const FlatFilePos& pos, size_t add_size, bool& out_
             if (file) {
                 LogDebug(BCLog::VALIDATION, "Pre-allocating up to position 0x%x in %s%05u.dat\n", new_size, m_prefix, pos.nFile);
                 AllocateFileRange(file, pos.nPos, inc_size);
-                fclose(file);
+                if (fclose(file) != 0) {
+                    LogError("Cannot close file %s%05u.dat after extending it with %u bytes", m_prefix, pos.nFile, new_size);
+                    return 0;
+                }
                 return inc_size;
             }
         } else {
@@ -86,17 +91,24 @@ bool FlatFileSeq::Flush(const FlatFilePos& pos, bool finalize) const
         return false;
     }
     if (finalize && !TruncateFile(file, pos.nPos)) {
-        fclose(file);
         LogError("%s: failed to truncate file %d\n", __func__, pos.nFile);
+        if (fclose(file) != 0) {
+            LogError("Failed to close file %d", pos.nFile);
+        }
         return false;
     }
     if (!FileCommit(file)) {
-        fclose(file);
         LogError("%s: failed to commit file %d\n", __func__, pos.nFile);
+        if (fclose(file) != 0) {
+            LogError("Failed to close file %d", pos.nFile);
+        }
         return false;
     }
     DirectoryCommit(m_dir);
 
-    fclose(file);
+    if (fclose(file) != 0) {
+        LogError("Failed to close file %d after flush", pos.nFile);
+        return false;
+    }
     return true;
 }
diff --git a/src/node/blockstorage.cpp b/src/node/blockstorage.cpp
index 557ddaceab..81f42827d9 100644
--- a/src/node/blockstorage.cpp
+++ b/src/node/blockstorage.cpp
@@ -961,6 +961,7 @@ bool BlockManager::WriteBlockUndo(const CBlockUndo& blockundo, BlockValidationSt
                 // Write undo data & checksum
                 fileout << blockundo << hasher.GetHash();
             }
+            // BufferedWriter will flush pending data to file when fileout goes out of scope.
         }
 
         // Make sure that the file is closed before we call `FlushUndoFile`.
diff --git a/src/node/mempool_persist.cpp b/src/node/mempool_persist.cpp
index 2551a43166..d4d9263973 100644
--- a/src/node/mempool_persist.cpp
+++ b/src/node/mempool_persist.cpp
@@ -201,8 +201,10 @@ bool DumpMempool(const CTxMemPool& pool, const fs::path& dump_path, FopenFn mock
         LogInfo("Writing %d unbroadcast transactions to file.\n", unbroadcast_txids.size());
         file << unbroadcast_txids;
 
-        if (!skip_file_commit && !file.Commit())
+        if (!skip_file_commit && !file.Commit()) {
+            (void)file.fclose();
             throw std::runtime_error("Commit failed");
+        }
         if (file.fclose() != 0) {
             throw std::runtime_error(
                 strprintf("Error closing %s: %s", fs::PathToString(file_fspath), SysErrorString(errno)));
diff --git a/src/rpc/blockchain.cpp b/src/rpc/blockchain.cpp
index 4f5a6b8434..be953cefde 100644
--- a/src/rpc/blockchain.cpp
+++ b/src/rpc/blockchain.cpp
@@ -83,7 +83,7 @@ UniValue WriteUTXOSnapshot(
     CCoinsViewCursor* pcursor,
     CCoinsStats* maybe_stats,
     const CBlockIndex* tip,
-    AutoFile& afile,
+    AutoFile&& afile,
     const fs::path& path,
     const fs::path& temppath,
     const std::function<void()>& interruption_point = {});
@@ -3115,7 +3115,14 @@ static RPCHelpMan dumptxoutset()
         }
     }
 
-    UniValue result = WriteUTXOSnapshot(*chainstate, cursor.get(), &stats, tip, afile, path, temppath, node.rpc_interruption_point);
+    UniValue result = WriteUTXOSnapshot(*chainstate,
+                                        cursor.get(),
+                                        &stats,
+                                        tip,
+                                        std::move(afile),
+                                        path,
+                                        temppath,
+                                        node.rpc_interruption_point);
     fs::rename(temppath, path);
 
     result.pushKV("path", path.utf8string());
@@ -3167,7 +3174,7 @@ UniValue WriteUTXOSnapshot(
     CCoinsViewCursor* pcursor,
     CCoinsStats* maybe_stats,
     const CBlockIndex* tip,
-    AutoFile& afile,
+    AutoFile&& afile,
     const fs::path& path,
     const fs::path& temppath,
     const std::function<void()>& interruption_point)
@@ -3244,12 +3251,19 @@ UniValue WriteUTXOSnapshot(
 UniValue CreateUTXOSnapshot(
     node::NodeContext& node,
     Chainstate& chainstate,
-    AutoFile& afile,
+    AutoFile&& afile,
     const fs::path& path,
     const fs::path& tmppath)
 {
     auto [cursor, stats, tip]{WITH_LOCK(::cs_main, return PrepareUTXOSnapshot(chainstate, node.rpc_interruption_point))};
-    return WriteUTXOSnapshot(chainstate, cursor.get(), &stats, tip, afile, path, tmppath, node.rpc_interruption_point);
+    return WriteUTXOSnapshot(chainstate,
+                             cursor.get(),
+                             &stats,
+                             tip,
+                             std::move(afile),
+                             path,
+                             tmppath,
+                             node.rpc_interruption_point);
 }
 
 static RPCHelpMan loadtxoutset()
diff --git a/src/rpc/blockchain.h b/src/rpc/blockchain.h
index b42fe96fd3..0e42bed947 100644
--- a/src/rpc/blockchain.h
+++ b/src/rpc/blockchain.h
@@ -51,7 +51,7 @@ void CalculatePercentilesByWeight(CAmount result[NUM_GETBLOCKSTATS_PERCENTILES],
 UniValue CreateUTXOSnapshot(
     node::NodeContext& node,
     Chainstate& chainstate,
-    AutoFile& afile,
+    AutoFile&& afile,
     const fs::path& path,
     const fs::path& tmppath);
 
diff --git a/src/streams.cpp b/src/streams.cpp
index 3c46530990..2c873e50d3 100644
--- a/src/streams.cpp
+++ b/src/streams.cpp
@@ -119,6 +119,7 @@ bool AutoFile::Commit()
 
 bool AutoFile::Truncate(unsigned size)
 {
+    m_was_written = true;
     return ::TruncateFile(m_file, size);
 }
 
diff --git a/src/test/util/chainstate.h b/src/test/util/chainstate.h
index eebf504d98..d5100d15b1 100644
--- a/src/test/util/chainstate.h
+++ b/src/test/util/chainstate.h
@@ -47,8 +47,11 @@ CreateAndActivateUTXOSnapshot(
     FILE* outfile{fsbridge::fopen(snapshot_path, "wb")};
     AutoFile auto_outfile{outfile};
 
-    UniValue result = CreateUTXOSnapshot(
-        node, node.chainman->ActiveChainstate(), auto_outfile, snapshot_path, snapshot_path); // Will close auto_outfile.
+    UniValue result = CreateUTXOSnapshot(node,
+                                         node.chainman->ActiveChainstate(),
+                                         std::move(auto_outfile), // Will close auto_outfile.
+                                         snapshot_path,
+                                         snapshot_path);
     LogPrintf(
         "Wrote UTXO snapshot to %s: %s\n", fs::PathToString(snapshot_path.make_preferred()), result.write());
 
```

</details>

I have checked remaining explicit closes, couldn't find any other one that was an `AutoFile`, but was wondering if we should maybe check these as well:
* https://github.com/bitcoin/bitcoin/blob/24e5fd3bedcebacbc10f0449be61be636b77dd79/src/util/fs_helpers.cpp#L54
* https://github.com/bitcoin/bitcoin/blob/24e5fd3bedcebacbc10f0449be61be636b77dd79/src/util/fs_helpers.cpp#L134
* https://github.com/bitcoin/bitcoin/blob/00e9b97f37e0bdf4c647236838c10b68b7ad5be3/src/util/readwritefile.cpp#L33

I have also rebased the PR locally and ran all tests.

vasild commented:
`f8a7580a08...c10e382d2a`: address suggestions

hodlinator reviewed (approved):
re-ACK c10e382d2a3b76b70ebb8a4eb5cd99fc9f14d702

Only minor changes in response to feedback since first ACK (https://github.com/bitcoin/bitcoin/pull/29307#pullrequestreview-2924706520)

hodlinator commented on src/addrdb.cpp:
```diff
@@ -61,25 +62,29 @@ bool SerializeFileDB(const std::string& prefix, const fs::path& path, const Data
         return false;
     }
-    fileout.fclose();
+    if (fileout.fclose() != 0) {
+        const int errno_save{errno};
+        remove(pathTmp);
+        LogError("Failed to close file %s after commit: %s", fs::PathToString(pathTmp), SysErrorString(errno_save));
```
Agree, would make sense to set `m_was_written=false` when we return `true` from `AutoFile::Commit()`. Maybe rename `m_was_written` to `m_pending_output`. Then this last `if`-block could be removed.

hodlinator replied:
If `Commit()` fails and (while failing) were to close the file internally as suggested, it should probably throw an exception after that so calling code doesn't try to continue using a closed `AutoFile`.

An alternative would be to have a `AutoFile::CommitAndClose()`-function that always closes regardless of errors and  clearly signals that calling code does not expect to continue using the `AutoFile`?

vasild replied:
I think the current `Commit()` is good as it is. It does what it promises - flushes the file data (and checks that the flush succeeded). However a subsequent close is not guaranteed to succeed if commit succeeded before that because close does more than flush.

hodlinator reviewed (commented):
> ~I am puzzled 😕 🤕~ edit: now it makes sense, `DirectoryCommit()` syncs the directory itself, not the file, I was blind. So that `fclose()` better be checked for an error. This was added in #14501

Might need to consider networked FS directory behavior when adding errors (https://github.com/bitcoin/bitcoin/pull/14501#discussion_r546662763)?

achow101 commented:
ACK c10e382d2a3b76b70ebb8a4eb5cd99fc9f14d702

fanquake commented:
@vasild Are you planning on responding to the (remaining) outstanding comments here? Or opening a followup?

vasild commented:
I prefer to always check whether a syscall failed even if it _seems_ that it should always succeed. In this particular case, the _"seems"_ is not very strong because `close` is doing [more things](https://github.com/bitcoin/bitcoin/pull/29307#discussion_r2161473468) than `flush` or `truncate` and those things could fail. So I think the current code is solid. I do not see a value in removing a check whether `close` failed.
